### ISSUE TITLE
LSTM Returning Questionable Results

### ISSUE THREAD

**Issue by farverio** [BrainJS/brain.js#271]
![Alt Text](https://media.giphy.com/media/12ZXZv7aoviBgI/giphy.gif)

## *What* is wrong?
I have trained an LSTM model but am getting poor results when testing strings (potentially two issues). The first is that results that one would presume to be negative instead are showing positive. Secondly, there are inputs tested with net.run() that don't return any result at all.

## *Where* does it happen?
I am running a linux machine using node. It occurs after training the data when loading that data with net.fromJSON() and running future examples with net.run().

## *How* do we replicate the issue?

I have included sample data and code here: https://jsfiddle.net/farverio/z378bnoa/6/

## *How* important is this (1-5)?
In the greater scheme of things, probably a 0-1 as its for a school project not a production environment but it feels like a 3-4 today and any help would still be appreciated :)

## Expected behavior (i.e. solution)
With any net.run(testString), I would presume the testString input is coerced to a positive or negative output (I only have two outputs).

Secondly, I'd have thought that the results might be more accurate. In the domain example, I am categorizing the interest of sample email responses as 'interested' or 'not interested'. When I type the exact copy of a not interested input that I used to train ("No interest."), it successfully returns not interested. However, when I remove the period at the end of the input, ("No interest"), it instead returns interested.

Is this normal behavior? How should I fix to produce more meaningful results?

## Other Comments
I tried to follow documentation pretty closely so am at a loss about both types of errors. Any help in resolving this would be greatly appreciated!!

**Comment by robertleeplummerjr** [comment]
Its a little late, I'll take a look in the morning, but the first thing that stands out is `iterations: 100`.  Crank that up to `iterations: 5000` or even `iterations: 10000`, and your results will vastly improve.

**Comment by farverio** [comment]
Thanks @robertleeplummerjr . Was looking over a lot of previous posts from you and hoping you'd reply. Appreciate you taking some time out of the holiday weekend to check things out :)

**Comment by robertleeplummerjr** [comment]
After playing with it a bit, I think the problem is just lack of data and overfitting.  If we look at your training set, the following have punctuation:
```
  {"input":"Thank you but we are not interested at this time.", "output":"Not Interested"},
  {"input":"No interest.", "output":"Not Interested"},
  {"input":"Dont contact me further!!", "output":"Not Interested"},
  {"input":"Please remove me from your list !!!!", "output":"Not Interested"},
  {"input":"No interest please remove us.", "output":"Not Interested"},
  {"input":"Please tell me more!", "output":"Interested"},
  {"input":"Are you free for a call?", "output":"Interested"},
  {"input":"How do I sign up?", "output":"Interested"},
  {"input":"Let's set up an introductory meeting.", "output":"Interested"}
```

The following to not have punctuation:
```
  {"input":"Not interested", "output":"Not Interested"},
  {"input":"No thank you", "output":"Not Interested"},
  {"input":"Please stop emailing", "output":"Not Interested"},
  {"input":"For now I am not interested", "output":"Not Interested"},
  {"input":"I'd like to hear more", "output":"Interested"},
  {"input":"Busy this week but let's schedule some time next week", "output":"Interested"},
  {"input":"I am open for a quick 10 minute call", "output":"Interested"},
  {"input":"What is the total cost", "output":"Interested"},
  {"input":"I forwarded your email to our CEO - he'll follow up", "output":"Interested"},
```

I don't think there is enough data yet for it to form a solid basis for a reply, but I could be wrong.  There are three sentences that end with period, and possibly that is having causing it to weigh towards "NotInterested".  I'm playing around with a deeper net to see if I can decouple this problem.

**Comment by robertleeplummerjr** [comment]
I cranked up the hidden layers, put iterations at 2k, errorThresh at 1.1, and added another input, it seemed to give a better result: https://jsfiddle.net/robertleeplummerjr/z378bnoa/29/

So in short, a deeper net and adjusting training settings should put you where you want to be.

**Comment by farverio** [comment]
Thanks for the reply! Can you help me understand how the hidden layers add more value? That just means essentially that it's testing more things throughout the iterations to find an "optimal" solution, right?

I'm running a test with actual cleansed data (not the test inputs), and am currently at 4100 iterations and running since I started after your last message yesterday. Training error has been at "Infinity" for some time now, so seems like it's struggling to find correlations?

Appreciate the continued help -- excited to be learning with Brain.js! Thanks again for all your help with the project!!

**Comment by robertleeplummerjr** [comment]
Hidden layers are like abstractions of ideas.  The more hidden layers, the more the next can abstract.  I think the net doesn't understand what you are wanting it to do.  Try giving it more layers, (like in the example I included) and see if that helps.
